cmd/build/v2/checker/sequential.go

```diff
@@ -0,0 +1,185 @@
+		tagger:               tagger,
+		imageCacheChecker:    imageCacheChecker,
+		metadataCollector:    metadataCollector,
+		ioCtrl:               ioCtrl,
+		serviceEnvVarsSetter: serviceEnvVarsSetter,
+	}
+}
+
+func (s *SequentialCheckStrategy) CheckServicesCache(ctx context.Context, manifestName string, buildManifest build.ManifestBuild, svcsToBuild []string) (cachedSvcs []string, notCachedSvcs []string, err error) {
+	// svcsToBuild is already ordered by dependencies (from DAG.Ordered())
+	// so we can optimize by stopping the check as soon as we find a service that's not cached.
+	// All subsequent services that depend on it will also need to be rebuilt.
+
+	dependantMap := make(map[string][]string)
+	// Build dependantMap: for each service, track which services depend on it
+	for svcName, buildInfo := range buildManifest {
+		for _, dep := range buildInfo.DependsOn {
+			dependantMap[dep] = append(dependantMap[dep], svcName)
+		}
+	}
+
+	// Track processed services to avoid duplicates
+	processed := make(map[string]bool)
+
+	for _, svc := range svcsToBuild {
+		// Skip if already processed
+		if processed[svc] {
+			continue
+		}
+
+		meta := s.metadataCollector.GetMetadata(svc)
+		start := time.Now()
+		buildInfo := buildManifest[svc]
+		buildHash := s.smartBuildCtrl.GetBuildHash(buildInfo, svc)
+
+		isCached, _, err := s.imageCacheChecker.IsCached(ctx, manifestName, buildInfo.Image, buildHash, svc)
+		if err != nil {
+			return cachedSvcs, notCachedSvcs, err
+		}
+		meta.CacheHitDuration = time.Since(start)
+
+		processed[svc] = true
+
+		if isCached {
+			meta.CacheHit = true
+			cachedSvcs = append(cachedSvcs, svc)
+		} else {
+			meta.CacheHit = false
+			notCachedSvcs = append(notCachedSvcs, svc)
+
+			// Recursively add all dependant services to notCached without checking cache
+			notCachedSvcs = s.addDependantsToNotCached(svc, dependantMap, processed, notCachedSvcs)
+		}
+	}
+	return cachedSvcs, notCachedSvcs, nil
+}
+
+// addDependantsToNotCached recursively adds all dependant services to notCached
+// This ensures that when a service is not cached, all services that depend on it
+// (directly or indirectly) are also marked as not cached without checking cache
+func (s *SequentialCheckStrategy) addDependantsToNotCached(svc string, dependantMap map[string][]string, processed map[string]bool, notCachedSvcs []string) []string {
+	if dependants, exists := dependantMap[svc]; exists {
+		for _, dependant := range dependants {
+			if !processed[dependant] {
+				processed[dependant] = true
+				notCachedSvcs = append(notCachedSvcs, dependant)
+				// Set metadata for dependant services
+				dependantMeta := s.metadataCollector.GetMetadata(dependant)
+				dependantMeta.CacheHit = false
+				dependantMeta.CacheHitDuration = 0 // No cache check performed
+
+				// Recursively add dependants of this dependant
+				notCachedSvcs = s.addDependantsToNotCached(dependant, dependantMap, processed, notCachedSvcs)
+			}
```

Comment:
Nit: it is easier to read and follow if we do an early break and we leave the main logic without one less level of indentation, specially taking into account that it would be the 3rd level.
```suggestion
			if processed[dependant] {
			    continue
			}
			processed[dependant] = true
			notCachedSvcs = append(notCachedSvcs, dependant)
			// Set metadata for dependant services
			dependantMeta := s.metadataCollector.GetMetadata(dependant)
			dependantMeta.CacheHit = false
			dependantMeta.CacheHitDuration = 0 // No cache check performed

			// Recursively add dependants of this dependant
			notCachedSvcs = s.addDependantsToNotCached(dependant, dependantMap, processed, notCachedSvcs)
			
```